R/thumbnail.R

```diff
@@ -0,0 +1,327 @@
+#' Get content item thumbnail
+#'
+#' Download the thumbnail for a content item on Connect to a file on your
+#' computer.
+#'
+#' @param content A content item.
+#' @param path Optional. A path to a file used to write the thumbnail image. If
+#' no path is provided, a temporary file with the correct file extension is
+#' created. If the content item does not have a thumbnail, returns `NA`.
+#' 
+#' @returns The path to the image file.
+#' 
+#' @examples
+#' \dontrun{
+#' client <- connect()
+#' item <- content_item(client, "8f37d6e0-3395-4a2c-aa6a-d7f2fe1babd0")
+#' thumbnail <- get_thumbnail(item)
+#' }
+#'
+#' @family thumbnail functions
+#' @family content functions
+#' @export
+get_thumbnail <- function(content, path = NULL) {
+  validate_R6_class(content, "Content")
+  guid <- content$content$guid
+  con <- content$connect
+
+  # Connect 2024.09.0 introduced public endpoints for content thumbnails. We
+  # prefer those, falling back to the unversioned endpoints if unavailable.
+  res <- con$GET(
+    url = con$server_url("content", guid, "__thumbnail__"),
+    parser = NULL
+  )
+  if (httr::status_code(res) == 404) {
+    res <- con$GET(
+      unversioned_url("applications", guid, "image"),
+      parser = NULL
+    )
+  }
+  con$raise_error(res)
+
+  if (httr::status_code(res) == 204) {
+    return(NA_character_)
+  }
+
+  # Guess file extension
+  if (is.null(path)) {
+    ct <- httr::headers(res)$`content-type`
+    if (grepl("^image/", ct)) {
+      # Just strip off 'image/'
+      ext <- substr(ct, 7, nchar(ct))
+      path <- tempfile(pattern = "content_image_", fileext = paste0(".", ext))
+    } else {
+      # Try png
+      warning(glue::glue("Could not infer file extension from content type: {ct}. Using '.png'"))
+      path <- tempfile(pattern = "content_image_", fileext = ".png")
+    }
```

Comment:
Could we do something like `ext <- names(mime::mimemap[mime::mimemap == ct])` using the map directly and so we aren't futzing with a parser here.

Separately / additionally: any reason to assume `png` if we can't figure out what the type is? That seems unhelpful no?

Reply:
This code is taken directly from the old implementation. Thinking about the default-PNG assumption — perhaps it's there for a reason, but I'm not sure.

I think it would be reasonable to use `mime::mimemap`. @nealrichardson and I were discussing this earlier (see https://github.com/rstudio/connectapi/pull/303#discussion_r1775910845). The `mimemap` conversion would be basically the same as just stripping the `image/`, except that there would be multiple matches, e.g.:

```
jpeg = "image/jpeg",
jpg = "image/jpeg",
jpe = "image/jpeg",
```

so, `ext <- names(mime::mimemap[mime::mimemap == ct])[1]`.

I don't personally find that much easier to comprehend than `substr(ct, 7, nchar(ct))`. But I guess if someone was using an SVG (is that even possible?) it'd be better to use the map, so yeah, I'm convinced.